src/terminal/adapter/MacroBuffer.cpp

```diff
@@ -0,0 +1,271 @@
+    // digits of each hex pair. But we also need to deal with repeat sequences,
+    // which start with `!`, followed by a numeric repeat count, and then a
+    // range of hex pairs between two `;` characters. When parsing the repeat
+    // count, we use the ExpectingRepeatCount state, but when parsing the hex
+    // pairs of the repeat, we just use the regular ExpectingHexDigit states.
+
+    auto success = true;
+    switch (_parseState)
+    {
+    case State::ExpectingText:
+        success = _appendToActiveMacro(ch);
+        break;
+    case State::ExpectingHexDigit:
+        if (_decodeHexDigit(ch))
+        {
+            _parseState = State::ExpectingSecondHexDigit;
+        }
+        else if (ch == L'!' && !_repeatPending)
+        {
+            _parseState = State::ExpectingRepeatCount;
+            _repeatCount = 0;
+        }
+        else if (ch == L';' && _repeatPending)
+        {
+            success = _applyPendingRepeat();
+        }
+        else
+        {
+            success = false;
+        }
+        break;
+    case State::ExpectingSecondHexDigit:
+        success = _decodeHexDigit(ch) && _appendToActiveMacro(_decodedChar);
+        _decodedChar = 0;
+        _parseState = State::ExpectingHexDigit;
+        break;
+    case State::ExpectingRepeatCount:
+        if (ch >= L'0' && ch <= L'9')
+        {
+            _repeatCount = _repeatCount * 10 + (ch - L'0');
+            _repeatCount = std::min<size_t>(_repeatCount, MAX_PARAMETER_VALUE);
+        }
+        else if (ch == L';')
+        {
+            _repeatPending = true;
+            _repeatStart = _activeMacro().length();
+            _parseState = State::ExpectingHexDigit;
+        }
+        else
+        {
+            success = false;
+        }
+        break;
+    default:
+        success = false;
+        break;
+    }
+
+    // If there is an error in the definition, clear everything received so far.
+    if (!success)
+    {
+        _deleteMacro(_activeMacro());
+    }
+    return success;
+}
+
+bool MacroBuffer::_decodeHexDigit(const wchar_t ch) noexcept
+{
+    _decodedChar <<= 4;
+    if (ch >= L'0' && ch <= L'9')
+    {
+        _decodedChar += (ch - L'0');
+        return true;
+    }
+    else if (ch >= L'A' && ch <= L'F')
+    {
+        _decodedChar += (ch - L'A' + 10);
+        return true;
+    }
+    else if (ch >= L'a' && ch <= L'f')
+    {
+        _decodedChar += (ch - L'a' + 10);
+        return true;
+    }
+    return false;
+}
+
+bool MacroBuffer::_appendToActiveMacro(const wchar_t ch)
+{
+    if (GetSpaceAvailable() > 0)
+    {
+        _activeMacro().push_back(ch);
+        _spaceUsed++;
+        return true;
+    }
+    return false;
+}
+
+std::wstring& MacroBuffer::_activeMacro() noexcept
+{
+    return til::at(_macros, _activeMacroId);
+}
+
+void MacroBuffer::_deleteMacro(std::wstring& macro) noexcept
+{
+    _spaceUsed -= macro.length();
+    std::wstring{}.swap(macro);

```

Comment:
Would `macro = {};` work here?

Reply:
I may be wrong, but I was under the impression that that wouldn't actually free up the memory.

Reply:
No that should for sure work. `.clear()` doesn't free memory, but in this case `{}` will construct a brand new instance of `std::wstring` and then move-assign it to `macro`, whose contents will be entirely replaced by the new instance.

Reply:
I've just tried this now and that doesn't seem to be the case. I created a 200K macro, and then deleted it, logging `macro.capacity()` before and after the delete. With the `macro = {}` the capacity remained at 200K. With `std::wstring{}.swap(macro)`, it shrunk down to 7 (I'm assuming that's the default). That suggested to me that `macro = {}` is not actually freeing the memory. Or have I misunderstood the meaning of `capacity`?

Reply:
Actually: https://godbolt.org/z/PMMq85cE7
So anyways, C++ is a simple language, that's easy to understand... /s 😑

It's because `std::basic_string` got a
```cpp
constexpr basic_string& operator=(std::initializer_list<CharT> ilist);
```